File: 005-2026-BI-INTELIGENCIA-NEGOCIO/docker/vanna/ollama_client.py
"""Cliente HTTP hacia Ollama · candados 2 (temp0/seed42/structured output)
· 3 (keep_alive) · 5 (metrics para el jurado)."""
from __future__ import annotations

import json
import logging
import os
from dataclasses import dataclass, field
from typing import Any

import httpx

log = logging.getLogger("ollama_client")

DEFAULT_TIMEOUT_S = 90.0
DEFAULT_KEEP_ALIVE_H = int(os.getenv("KEEP_ALIVE_H", "24"))


@dataclass
class OllamaMetrics:
    latencia_ms: int
    prompt_tokens: int | None = None
    response_tokens: int | None = None
    load_duration_ms: int | None = None


@dataclass
class OllamaResultado:
    data: dict[str, Any]
    raw_response: str
    metrics: OllamaMetrics
    metadatos: dict[str, Any] = field(default_factory=dict)


class OllamaError(RuntimeError):
    """Cualquier fallo hablando con Ollama (timeout, HTTP, JSON inválido)."""


async def generar_estructurado(
    modelo: str,
    prompt: str,
    schema: dict[str, Any],
    system: str | None = None,
    options: dict[str, Any] | None = None,
    keep_alive_h: int | None = None,
    base_url: str | None = None,
    client: httpx.AsyncClient | None = None,
    timeout_s: float = DEFAULT_TIMEOUT_S,
) -> OllamaResultado:
    """Llama POST /api/generate de Ollama con structured output.

    Candado 2: options siempre lleva temperature=0 y seed=42 (defaults duros).
    Candado 3: keep_alive obligatorio para preservar modelos calientes.
    """
    base = base_url or os.getenv("OLLAMA_BASE_URL", "http://100.91.87.86:11435")
    opts = {"temperature": 0, "seed": 42}
    if options:
        opts.update(options)
    keep_alive = f"{keep_alive_h if keep_alive_h is not None else DEFAULT_KEEP_ALIVE_H}h"
    body: dict[str, Any] = {
        "model": modelo,
        "prompt": prompt,
        "stream": False,
        "format": schema,
        "options": opts,
        "keep_alive": keep_alive,
    }
    if system:
        body["system"] = system

    close_after = client is None
    if client is None:
        client = httpx.AsyncClient(timeout=timeout_s)
    try:
        try:
            resp = await client.post(f"{base}/api/generate", json=body)
        except httpx.TimeoutException as e:
            log.warning("ollama_timeout modelo=%s err=%s", modelo, e)
            raise OllamaError(f"timeout:{e}") from e
        except httpx.HTTPError as e:
            log.warning("ollama_http_error modelo=%s err=%s", modelo, e)
            raise OllamaError(f"http:{e}") from e

        if resp.status_code != 200:
            raise OllamaError(f"http_{resp.status_code}:{resp.text[:200]}")

        payload = resp.json()
        raw = payload.get("response", "")
        try:
            data = json.loads(raw) if isinstance(raw, str) else raw
        except json.JSONDecodeError as e:
            log.warning("ollama_bad_json modelo=%s raw=%s", modelo, raw[:200])
            raise OllamaError(f"bad_json:{e}") from e
        if not isinstance(data, dict):
            raise OllamaError("respuesta_no_es_objeto")

        metrics = OllamaMetrics(
            latencia_ms=int(payload.get("total_duration", 0) / 1_000_000)
            if isinstance(payload.get("total_duration"), int)
            else 0,
            prompt_tokens=payload.get("prompt_eval_count"),
            response_tokens=payload.get("eval_count"),
            load_duration_ms=int(payload.get("load_duration", 0) / 1_000_000)
            if isinstance(payload.get("load_duration"), int)
            else None,
        )
        return OllamaResultado(data=data, raw_response=raw, metrics=metrics)
    finally:
        if close_after:
            await client.aclose()


async def listar_modelos(base_url: str | None = None, timeout_s: float = 3.0) -> list[str]:
    base = base_url or os.getenv("OLLAMA_BASE_URL", "http://100.91.87.86:11435")
    async with httpx.AsyncClient(timeout=timeout_s) as client:
        try:
            resp = await client.get(f"{base}/api/tags")
        except httpx.HTTPError as e:
            raise OllamaError(f"tags_http:{e}") from e
    if resp.status_code != 200:
        raise OllamaError(f"tags_status_{resp.status_code}")
    data = resp.json()
    modelos = data.get("models", []) if isinstance(data, dict) else []
    return [m.get("name", "") for m in modelos if isinstance(m, dict) and m.get("name")]

File: 006-2026-BI-INTELIGENCIA-NEGOCIO/src/lib/ai/ollama-client.ts
// src/lib/ai/ollama-client.ts · Cliente HTTP para Ollama (Mac Studio vía Tailscale)
// Producto 006 · BI v2 · Admin IA
// Adaptado de PI (002 · src/lib/ai/ollama-client.ts, referencia SOLO LECTURA).
// Logging explícito de cada llamada (modelo, tokens, latencia, éxito/fracaso)
// con prefijo [OLLAMA] — el 006 no tiene lib/logger: console.error/warn.
// R2: antes de CUALQUIER generación se exige URL local (assertUrlOllamaLocal).

import { assertUrlOllamaLocal, getOllamaBaseUrl, getOllamaTimeoutMs } from "./ollama-config";

// Ollama devuelve duraciones en nanosegundos; normalizamos a milisegundos
// para mantener consistencia con latenciaMs.
function nsToMs(ns: number | undefined): number | null {
    if (ns === undefined || ns === null) return null;
    return ns / 1_000_000;
}

interface OllamaResponse {
    model: string;
    created_at: string;
    response: string;
    thinking?: string;
    done: boolean;
    done_reason?: string;
    context?: number[];
    total_duration?: number;
    load_duration?: number;
    prompt_eval_count?: number;
    prompt_eval_duration?: number;
    eval_count?: number;
    eval_duration?: number;
}

export interface OllamaMetrics {
    modelo: string;
    latenciaMs: number;
    promptTokens: number | null;
    responseTokens: number | null;
    totalDuration: number | null;
    loadDuration: number | null;
}

export interface ResultadoPruebaModelo {
    respuesta: string;
    metrics: OllamaMetrics;
}

/**
 * Prueba un modelo con un prompt libre (playground Admin IA): POST
 * /api/generate con stream:false y opciones deterministas (temperature 0,
 * seed 42). Devuelve la respuesta en texto y las métricas de la llamada.
 *
 * Nota: modelos con thinking mode (Qwen3 y similares) pueden devolver el
 * texto en el campo `thinking` cuando `response` sale vacío — mismo
 * fallback que PI.
 */
export async function probarModelo(modelo: string, prompt: string): Promise<ResultadoPruebaModelo> {
    const inicio = Date.now();
    const baseUrl = await getOllamaBaseUrl();
    assertUrlOllamaLocal(baseUrl);
    const timeoutMs = await getOllamaTimeoutMs();

    const response = await fetch(`${baseUrl}/api/generate`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({
            model: modelo,
            prompt,
            stream: false,
            options: { temperature: 0, seed: 42 },
        }),
        signal: AbortSignal.timeout(timeoutMs),
    });

    const latenciaMs = Date.now() - inicio;

    if (!response.ok) {
        const errorText = await response.text().catch(() => "unknown");
        console.error(`[OLLAMA] ERROR probarModelo modelo=${modelo} status=${response.status} latencia=${latenciaMs}ms errorLen=${errorText.length}`);
        throw new Error(`Ollama HTTP ${response.status}: ${errorText}`);
    }

    const data = (await response.json()) as OllamaResponse;

    const metrics: OllamaMetrics = {
        modelo,
        latenciaMs,
        promptTokens: data.prompt_eval_count ?? null,
        responseTokens: data.eval_count ?? null,
        totalDuration: nsToMs(data.total_duration),
        loadDuration: nsToMs(data.load_duration),
    };

    const respuesta = data.response?.trim() || data.thinking?.trim() || "";

    console.warn(`[OLLAMA] OK probarModelo modelo=${modelo} latencia=${latenciaMs}ms promptTokens=${metrics.promptTokens} responseTokens=${metrics.responseTokens}`);

    return { respuesta, metrics };
}

/**
 * Llama a Ollama con structured output nativo (JSON Schema) usando
 * /api/generate (Fase 2 · motor NL→SQL). Requiere Ollama >= 0.29.0 y un
 * modelo que soporte `format` con schema. Misma lógica que PI: opciones
 * deterministas por defecto (temperature 0, seed 42), fallback al campo
 * `thinking` si `response` sale vacío, y error claro si el JSON no parsea
 * (se tira la salida; NO se rescata a la fuerza — candado 2).
 */
export async function llamarOllamaStructured<T>(
    modelo: string,
    prompt: string,
    schema: Record<string, unknown>,
    system?: string,
    options?: Record<string, unknown>
): Promise<{ data: T; rawResponse: string; metrics: OllamaMetrics }> {
    const startTime = Date.now();

    const body: Record<string, unknown> = {
        model: modelo,
        prompt,
        stream: false,
        format: schema,
        options: {
            temperature: 0,
            seed: 42,
            ...options,
        },
    };
    if (system) body.system = system;

    const ollamaBaseUrl = await getOllamaBaseUrl();
    assertUrlOllamaLocal(ollamaBaseUrl);
    const timeoutMs = await getOllamaTimeoutMs();
    const response = await fetch(`${ollamaBaseUrl}/api/generate`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(body),
        signal: AbortSignal.timeout(timeoutMs),
    });

    const latenciaMs = Date.now() - startTime;

    if (!response.ok) {
        const errorText = await response.text().catch(() => "unknown");
        console.error(`[OLLAMA] ERROR structured modelo=${modelo} status=${response.status} latencia=${latenciaMs}ms errorLen=${errorText.length}`);
        throw new Error(`Ollama HTTP ${response.status}: ${errorText}`);
    }

    const data = (await response.json()) as OllamaResponse;

    const metrics: OllamaMetrics = {
        modelo,
        latenciaMs,
        promptTokens: data.prompt_eval_count ?? null,
        responseTokens: data.eval_count ?? null,
        totalDuration: nsToMs(data.total_duration),
        loadDuration: nsToMs(data.load_duration),
    };

    const rawResponse = data.response?.trim() || data.thinking?.trim() || "";

    let parsed: T;
    try {
        parsed = JSON.parse(rawResponse) as T;
    } catch {
        console.error(`[OLLAMA] JSON inválido structured modelo=${modelo} latencia=${latenciaMs}ms responseLen=${rawResponse.length}`);
        throw new Error("Ollama devolvió JSON inválido a pesar del schema");
    }

    console.warn(`[OLLAMA] OK structured modelo=${modelo} latencia=${latenciaMs}ms promptTokens=${metrics.promptTokens} responseTokens=${metrics.responseTokens}`);

    return { data: parsed, rawResponse, metrics };
}
